Servicing CLR based DNX runtimes

moozzyk opened the pull request:
Allows to use a different runtime based on index.txt file

moozzyk commented:
/CC: @victorhurdugaci @muratg @davidfowl 

@blowdart - this will need security review

blowdart commented:
Can you get me some test results, and set up a meeting next week to go through the code once I get back? Thanks

davidfowl commented on src/dnx/servicing.cpp:
```diff
@@ -0,0 +1,94 @@
+        std::wstring get_runtime_path(const std::wstring& servicing_root_parent, TraceWriter& trace_writer)
+        {
+            auto servicing_root = utils::path_combine(servicing_root_parent, std::wstring(L"Microsoft DNX"));
+            auto servicing_manifest_path = utils::path_combine(servicing_root, std::wstring(L"Servicing\\index.txt"));
```
Shouldn;'t this be multiple path_combine calls?

moozzyk replied:
Does it make any difference (apart from copying the string twice)?

victorhurdugaci commented on src/dnx/pal.win32.cpp:
```diff
@@ -81,19 +83,56 @@ BOOL GetFullPath(LPCTSTR szPath, LPTSTR pszNormalizedPath)
+#endif
+        };
+
+        for (auto i = 0; i < sizeof(program_files_paths)/sizeof(wchar_t*); i++)
```
Do you really need a `for` loop here? There are just 2 elements

moozzyk replied:
Are you suggesting I should copy the code twice? (If so the answer is yes, I need a loop).

moozzyk replied:
On 32-bit windows ProgramFiles(x86) does not exist.

moozzyk replied:
We are now always checking `%ProgramFiles(x86)%` and only if it does not exist we will add  `%ProgramFiles%` to the list

victorhurdugaci commented on src/dnx/servicing.cpp:
```diff
@@ -0,0 +1,94 @@
+        {
+            if (line.compare(0, runtime_qualifier.length(), runtime_qualifier) == 0)
+            {
+                return dnx::utils::to_wstring(line.substr(runtime_qualifier.length()));
```
This code seems to pick the first matching entry in the index. It must always pick the latest because new patches are added at the end of the file

moozzyk replied:
I think when I talked to @glennc he said there would be only one entry for each source runtime. However if we panning only on appending entries what you are saying makes sense.

victorhurdugaci replied:
For nupkgs that's how it work. We always append new lines, never delete

moozzyk replied:
OK. I will confirm with @glennc and fix to take the last occurence.

moozzyk replied:
As per our conversation - there can be multiple entries for the same runtime. We will now take the last entry from the index.txt.

victorhurdugaci commented on src/dnx/pal.win32.cpp:
```diff
@@ -81,19 +83,56 @@ BOOL GetFullPath(LPCTSTR szPath, LPTSTR pszNormalizedPath)
+        if (runtime_path.length() > 0)
+        {
+            traceWriter.Write(std::wstring(L"Redirecting runtime to: ").append(runtime_path), true);
+            SetEnvironmentVariable(_T("DNX_DEFAULT_LIB"), runtime_path.c_str());
```
Where is this env variable used?

moozzyk replied:
In a few places but most notably in the app domain manager(s). It's originally set (here)[https://github.com/aspnet/dnx/blob/dev/src/dnx/dnx.cpp#L259] and updated for servicing to force loading managed assemblies from the redirected runtime.

victorhurdugaci commented on src/dnx/pal.win32.cpp:
```diff
@@ -81,19 +83,56 @@ BOOL GetFullPath(LPCTSTR szPath, LPTSTR pszNormalizedPath)
+namespace
+{
+    std::wstring get_runtime_path(TraceWriter& trace_writer)
+    {
```
The servicing index can be located in a different folder, other than program files, if you set the `DNX_SERVICING` env var. https://github.com/aspnet/dnx/blob/32358ec9b8970070ff77cbc5beef28c13e3ab09b/src/Microsoft.Framework.Runtime/Servicing/ServicingTable.cs#L31

moozzyk replied:
Must have missed (or forgotten) when reading the spec. Will fix.

moozzyk replied:
Added support for DNX_SERVICING. If the file does not exists or no entries are found in this file we will fall back and check the index.txt in the `%ProgramFiles%` if one exists.

moozzyk commented:
@victorhurdugaci - updated the PR with:
- added support for DNX_SERVICING
- added fallback mechanism to use %ProgramFiles% if no matching entries are found in the `%DNX_SERVICING\%index.txt` file
- we will now use the last matching entry from the index.txt file instead of using the first one
